### PR TITLE
add support for namespaces to WikiaView

### PR DESCRIPTION
**Pull request by hakubo**
This PR adds support for namespaces to WikiaView, so we can have nirvana controllers also inside namespaces. Template path should still be build based on class name but without namespace.

Example:
$this->renderView('PageHeader/PageHeader', 'index');

Alternative idea:
add a param to renderView - $namespace, so we could generate paths correctly
this would unnecessarily increase learning curve of how to render view in nirvana

Current approach mimics how you would use a class inside a namespace in PHP
but allows to have all templates inside ClassName/templates, folder

### CODE REVIEW

**Comment by WikiaTech** [comment]
SonarQube analysis reported 8 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [WikiaView.class.php#L161](https://github.com/Wikia/app/blob/2f44ceca9fd6f491cb1d5d8e888cdbda8a438a58/includes/wikia/nirvana/WikiaView.class.php#L161): Remove this commented out code. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS125)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L274](https://github.com/Wikia/app/blob/2f44ceca9fd6f491cb1d5d8e888cdbda8a438a58/includes/wikia/nirvana/WikiaView.class.php#L274): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L275](https://github.com/Wikia/app/blob/2f44ceca9fd6f491cb1d5d8e888cdbda8a438a58/includes/wikia/nirvana/WikiaView.class.php#L275): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L313](https://github.com/Wikia/app/blob/2f44ceca9fd6f491cb1d5d8e888cdbda8a438a58/includes/wikia/nirvana/WikiaView.class.php#L313): Remove the code after this "return". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1763)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L316](https://github.com/Wikia/app/blob/2f44ceca9fd6f491cb1d5d8e888cdbda8a438a58/includes/wikia/nirvana/WikiaView.class.php#L316): Reduce this "switch/case" number of lines from 21 to at most 10, for example by extracting code into function. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1151)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L335](https://github.com/Wikia/app/blob/2f44ceca9fd6f491cb1d5d8e888cdbda8a438a58/includes/wikia/nirvana/WikiaView.class.php#L335): Remove the code after this "return". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1763)
1. ![MINOR][MINOR] [WikiaView.class.php#L274](https://github.com/Wikia/app/blob/2f44ceca9fd6f491cb1d5d8e888cdbda8a438a58/includes/wikia/nirvana/WikiaView.class.php#L274): 2 statements were found on this line. Reformat the code to have only one statement per line. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS122)
1. ![MINOR][MINOR] [WikiaView.class.php#L275](https://github.com/Wikia/app/blob/2f44ceca9fd6f491cb1d5d8e888cdbda8a438a58/includes/wikia/nirvana/WikiaView.class.php#L275): 2 statements were found on this line. Reformat the code to have only one statement per line. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS122)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 8 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [WikiaView.class.php#L161](https://github.com/Wikia/app/blob/a30164381b15bd810dafda137c72ef43a11e7aa4/includes/wikia/nirvana/WikiaView.class.php#L161): Remove this commented out code. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS125)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L281](https://github.com/Wikia/app/blob/a30164381b15bd810dafda137c72ef43a11e7aa4/includes/wikia/nirvana/WikiaView.class.php#L281): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L282](https://github.com/Wikia/app/blob/a30164381b15bd810dafda137c72ef43a11e7aa4/includes/wikia/nirvana/WikiaView.class.php#L282): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L320](https://github.com/Wikia/app/blob/a30164381b15bd810dafda137c72ef43a11e7aa4/includes/wikia/nirvana/WikiaView.class.php#L320): Remove the code after this "return". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1763)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L323](https://github.com/Wikia/app/blob/a30164381b15bd810dafda137c72ef43a11e7aa4/includes/wikia/nirvana/WikiaView.class.php#L323): Reduce this "switch/case" number of lines from 21 to at most 10, for example by extracting code into function. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1151)
1. ![MAJOR][MAJOR] [WikiaView.class.php#L342](https://github.com/Wikia/app/blob/a30164381b15bd810dafda137c72ef43a11e7aa4/includes/wikia/nirvana/WikiaView.class.php#L342): Remove the code after this "return". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1763)
1. ![MINOR][MINOR] [WikiaView.class.php#L281](https://github.com/Wikia/app/blob/a30164381b15bd810dafda137c72ef43a11e7aa4/includes/wikia/nirvana/WikiaView.class.php#L281): 2 statements were found on this line. Reformat the code to have only one statement per line. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS122)
1. ![MINOR][MINOR] [WikiaView.class.php#L282](https://github.com/Wikia/app/blob/a30164381b15bd810dafda137c72ef43a11e7aa4/includes/wikia/nirvana/WikiaView.class.php#L282): 2 statements were found on this line. Reformat the code to have only one statement per line. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS122)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: includes/wikia/nirvana/WikiaView.class.php [hunk 1]
```diff
@@ -295,7 +316,6 @@ protected function renderHtml() {
 				wfProfileOut(__METHOD__);
 
 				return $result;
-				break;
 			case WikiaResponse::TEMPLATE_ENGINE_PHP:
 			default:
```
Comment (WikiaTech) [inline includes/wikia/nirvana/WikiaView.class.php:309]: MAJOR Reduce this "switch/case" number of lines from 20 to at most 10, for example by extracting code into function. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1151)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Review by rogatty — APPROVED** [review]
I worry that forcing the types might break some old code where the methods were used incorrectly. It should be caught by the tests, though. And it's better to fail loudly than do weird things silently, right? :wink:

**Review by latata — APPROVED** [review]

File: includes/wikia/nirvana/WikiaView.class.php [hunk 2]
```diff
@@ -178,23 +176,27 @@ private static function normalizeControllerClass( $controllerClass, &$controller
...
 	 *
-	 * @param string $controllerClass
+	 * @param string $controllerClassWithNamespace
 	 * @param string $methodName
 	 * @param string $controllerBaseName (optional) save cpu cycles by not
 	 *
 	 * @return array
 	 */
-	private function getTemplateOptions( $controllerClass, $methodName, $controllerBaseName = null ) {
+	private function getTemplateOptions( string $controllerClassWithNamespace, string $methodName, $controllerBaseName = null ): array {
```
Comment (kvas-damian) [inline includes/wikia/nirvana/WikiaView.class.php:177]: Can we add some unit tests covering this change?

**Comment by macbre** [comment]
Please update `WikiaViewTest` to cover new use case.

File: includes/wikia/nirvana/WikiaView.class.php [hunk 3]
```diff
@@ -250,13 +265,19 @@ private function getTemplateDir( $controllerClass ) {
...
 
-	public function __toString() {
+	public function __toString(): string {
 		try {
 			return $this->render();
 		} catch( Exception $e ) {
 			// php doesn't allow exceptions to be thrown inside __toString() so we need an extra try/catch block here
-			if ($this->response == null) return "WikiaView: response object was null rendering {$this->templatePath}";
-			if ($this->response->getException() == null) $this->response->setException($e);
+			if ($this->response == null) {
```
Comment (mszabo-wikia) [inline includes/wikia/nirvana/WikiaView.class.php]: Spacing

File: includes/wikia/nirvana/WikiaView.class.php [hunk 4]
```diff
@@ -250,13 +265,19 @@ private function getTemplateDir( $controllerClass ) {
...
+				return "WikiaView: response object was null rendering {$this->templatePath}";
+			}
+
+			if ($this->response->getException() == null) {
+				$this->response->setException($e);
```
Comment (mszabo-wikia) [inline includes/wikia/nirvana/WikiaView.class.php]: Let's add spacing here and use strict equality operator

File: includes/wikia/nirvana/WikiaView.class.php [hunk 5]
```diff
@@ -178,23 +176,27 @@ private static function normalizeControllerClass( $controllerClass, &$controller
...
+		$fromAnnotation = $this->getTemplateAnnotation( $controllerClassWithNamespace, $methodName );
 		if ( !empty( $fromAnnotation ) ) {
 			$templates[] = $fromAnnotation;
 		}
 
 		// Add variations on the controller name
 		if ( $controllerBaseName === null ) {
-			$controllerBaseName = F::app()->getBaseName( $controllerClass );
+			$controllerBaseName = $this->getClassName(
+				F::app()->getBaseName( $controllerClass )
```
Comment (mszabo-wikia) [inline includes/wikia/nirvana/WikiaView.class.php]: Here `$controllerClass` was already passed through `WikiaView::getClassName`, so the namespaces should be gone. Should we pass it again?

File: includes/wikia/nirvana/WikiaView.class.php [hunk 6]
```diff
@@ -204,7 +206,20 @@ private function getTemplateOptions( $controllerClass, $methodName, $controllerB
 		return $templates;
 	}
 
-	protected function getTemplateAnnotation( $controllerClass, $methodName ) {
+	private function getClassName( string $controllerClassWithNamespace ): string {
```
Comment (mszabo-wikia) [inline includes/wikia/nirvana/WikiaView.class.php:199]: What if we added this logic to `WikiaApp::getBaseName`? In that case we don't have to modify `WikiaView`.

**Review by mszabo-wikia — CHANGES_REQUESTED** [review]

**Review by macbre — CHANGES_REQUESTED** [review]
Unit tests, bitte :smile:

File: includes/wikia/nirvana/WikiaView.class.php (hunk 6 above)
Comment (hakubo) [reply includes/wikia/nirvana/WikiaView.class.php:199]: then it could not run the class, I need to remove namespace only to build template path

**Review by mszabo-wikia — APPROVED** [review]

**Review by kvas-damian — APPROVED** [review]

**Review by macbre — APPROVED** [review]

File: includes/wikia/tests/_fixtures/TestATestController.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace TestA;
+use \TestController as BaseTestController;
+use \AnotherTestController as BaseAnotherTestController;
+
+global $wgAutoloadClasses;
+$wgAutoloadClasses['TestA\TestController'] = dirname(__FILE__) . '/TestATestController.php';
+$wgAutoloadClasses['TestA\AnotherTestController'] = dirname(__FILE__) . '/TestATestController.php';
```
Comment (macbre) [inline includes/wikia/tests/_fixtures/TestATestController.php]: Let's use `__DIR__` instead

File: includes/wikia/tests/_fixtures/TestATestBTestController.php [hunk 8]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+namespace TestA\TestB;
+use \TestController as BaseTestController;
+use \AnotherTestController as BaseAnotherTestController;
+
+global $wgAutoloadClasses;
+$wgAutoloadClasses['TestA\TestB\TestController'] = dirname(__FILE__) . '/TestATestBTestController.php';
+$wgAutoloadClasses['TestA\TestB\AnotherTestController'] = dirname(__FILE__) . '/TestATestBTestController.php';
```
Comment (macbre) [inline includes/wikia/tests/_fixtures/TestATestBTestController.php]: Let's use `__DIR__` instead

**Pull request merged**